### PR TITLE
Adds paddle grid_sample for spatial transformation and paddle floor_mod

### PR DESCRIPTION
**Pull request by Angel-ica**
- [ ] #20198 

- [ ] #21377  

This is my first contribution to the project and I'm eager to learn and improve (especially with the grid_sample function as i found it challenging).  
Any feedback and guidance would be greatly appreciated.

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 1]
```diff
@@ -135,3 +136,169 @@ def affine_grid(theta, out_shape, align_corners=True):
...
+
+    Parameters:
+    - input: image to be transformed.
+    - grid: defines the transformation to be applied to the input tensor.
+    - mode: determines the interpolation method.
+    - padding_mode: controls how to handle out-of-bound values.
+
+    Returns:
+    Transformed tensor.
+    """
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:162]: we don't add docstrings to frontend functions

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 2]
```diff
@@ -135,3 +136,169 @@ def affine_grid(theta, out_shape, align_corners=True):
...
+
+    parameters:
+    - input: image that needs to be transformed.
+    - grid: the transformation grid.
+    - padding_mode: controls how to handle out-of-bound values.
+
+    Returns:
+    transformed values for each point in the output grid.
+    These transformed values represent the result of applying spatial transformation.
+    """
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:198]: please remove the docstring and rename the function to reflect it's private nature

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 3]
```diff
@@ -135,3 +136,169 @@ def affine_grid(theta, out_shape, align_corners=True):
...
+
+    Parameters:
+    - input: data that needs to be transformed.
+    - grid: defines the transformation to be applied to the input tensor.
+    - padding_mode: controls how to handle out-of-bounds values.
+
+    Returns:
+    The transformed values for each point in the output grid.
+    These values represent the result of applying spatial transformation.
+    """
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:244]: please remove the docstring and rename the function to reflect it's private nature

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py:1]: please take a look on how frontend tests are implemented, this implementation is not correct

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py
Comment (Angel-ica) [reply ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py:1]: Hi Mahmoud! Thank you for your feedback, I've removed all docstrings and reattempted the tests. 
Please review the new changes.

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 4]
```diff
@@ -135,3 +136,118 @@ def affine_grid(theta, out_shape, align_corners=True):
             base_grid[:, :, :, :, 3] = ivy.full((D, H, W), 1)
             grid = ivy.matmul(base_grid.view((N, D * H * W, 4)), theta.swapaxes(1, 2))
             return grid.view((N, D, H, W, 3))
+
+
+@to_ivy_arrays_and_back
+@with_unsupported_dtypes({"2.5.1 and below": ("float16", "bfloat16")}, "paddle")
+def grid_sample(input, grid, mode, padding_mode):
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:143]: the fronend function signature must match the native framework function

```suggestion
def grid_sample(
    x,
    grid,
    mode='bilinear',
    padding_mode='zeros',
    align_corners=True,
    name=None,
):
```

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 5]
```diff
@@ -135,3 +136,118 @@ def affine_grid(theta, out_shape, align_corners=True):
...
+            "The grid should be a 4D tensor with the last dim having size 2."
+        )  # size 2 representing (x,y) coordinates.
+
+    if mode == "nearest":
+        return grid_sample_nearest(input, grid, padding_mode)
+    elif mode == "bilinear":
+        return grid_sample_bilinear(input, grid, padding_mode)
+
+
+def grid_sample_nearest(input, grid, padding_mode):
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:164]: please rename to reflect its private nature

File: ivy/functional/frontends/paddle/nn/functional/vision.py [hunk 6]
```diff
@@ -135,3 +136,118 @@ def affine_grid(theta, out_shape, align_corners=True):
...
+    x_indices = ((grid_x + 1.0) * 0.5 * (in_width - 1)).astype("int32")
+    y_indices = ((grid_y + 1.0) * 0.5 * (in_height - 1)).astype("int32")
+    # Gather values from input using indices
+    gathered = input[:, :, y_indices, x_indices]
+
+    return gathered
+    # the shape of the gathered tensor is (batch, channels, out_height, out_width)
+
+
+def grid_sample_bilinear(input, grid, padding_mode):
```
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/nn/functional/vision.py:193]: rename

File: ivy/functional/frontends/paddle/tensor/math.py
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/paddle/tensor/math.py:1]: please revert all the code that is not related to your PR

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_vision.py:1]: you need to write the test for the main function, not the two helpers

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**